File: leetcode/82_Remove_Duplicates_from_Sorted_List_II.py
import unittest
from typing import Optional, List

class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next

    def __eq__(self, other) -> bool:
        if not isinstance(other, self.__class__):
            return False
        return self.val == other.val and self.next == other.next

    def __repr__(self) -> str:
        vals = []
        current = self
        while current is not None:
            vals.append(current.val)
            current = current.next
        return vals.__repr__()

class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy_head = ListNode(None, head)
        current = dummy_head
        delete_val = None
        while current.next is not None:
            if current.next.val == delete_val:
                current.next = current.next.next
            elif current.next.next is not None and current.next.val == current.next.next.val:
                delete_val = current.next.val
                current.next = current.next.next
            else:
                current = current.next
        return dummy_head.next

class TestDeleteDuplicates(unittest.TestCase):

    @classmethod
    def list_to_listnode(cls, lis: List[int]) -> Optional[ListNode]:
        if len(lis) > 0:
            return ListNode(lis[0], cls.list_to_listnode(lis[1:]))
        return None


    def test_1(self) -> None:
        self.assertEqual(
            Solution().deleteDuplicates(self.list_to_listnode([1,2,3,3,4,4,5])),
            self.list_to_listnode([1,2,5])
        )

    def test_2(self) -> None:
        self.assertEqual(
            Solution().deleteDuplicates(self.list_to_listnode([1,1,1,2,3])),
            self.list_to_listnode([2,3])
        )

File: leetcode/82_Remove_Duplicates_from_Sorted_List_II.rs
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val
        }
    }
}

struct Solution {}

impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut head = head;
        let mut current = &mut head;
        while current.is_some() {
            let cur_value = current.as_ref().unwrap().val;
            if current.as_ref().unwrap().next.is_some() && cur_value == current.as_ref().unwrap().next.as_ref().unwrap().val {
                while current.is_some() && cur_value == current.as_ref().unwrap().val {
                    *current = current.as_mut().unwrap().next.take();
                }
            } else {
                current = &mut current.as_mut().unwrap().next;
            }
        }
        head
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn array_into_list(slice: &[i32]) -> Option<Box<ListNode>> {
        if slice.is_empty() {
            None
        } else {
            let mut node = Box::new(ListNode::new(slice[0]));
            node.next = array_into_list(&slice[1..]);
            Some(node)
        }
    }
    
    #[test]
    fn test_1() {
        assert_eq!(
            Solution::delete_duplicates(array_into_list(&[1,2,3,3,4,4,5])), 
            array_into_list(&[1,2,5])
        );
    }

    #[test]
    fn test_2() {
        assert_eq!(
            Solution::delete_duplicates(array_into_list(&[1,1,1,2,3])), 
            array_into_list(&[2,3])
        );
    }
}